pkg/components/internal/testutil/jsonpath.go

```diff
@@ -15,70 +15,65 @@
 package testutil
 
 import (
+	"fmt"
 	"reflect"
+	"strings"
 	"testing"
 
+	"github.com/kinvolk/lokomotive/pkg/k8sutil"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-	yamlserializer "k8s.io/apimachinery/pkg/runtime/serializer/yaml"
 	"k8s.io/client-go/util/jsonpath"
 )
 
-// unstructredObj accepts a Kubernetes manifest in YAML format and returns an object of type
-// `unstructured.Unstructured`. This object has many methods that can be used by the consumer to
-// extract metadata from the Kubernetes manifest.
-func unstructredObj(t *testing.T, yamlObj string) *unstructured.Unstructured {
-	u := &unstructured.Unstructured{}
-
-	// Decode YAML into `unstructured.Unstructured`.
-	dec := yamlserializer.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
-	if _, _, err := dec.Decode([]byte(yamlObj), nil, u); err != nil {
-		t.Fatalf("Converting config to unstructured.Unstructured: %v", err)
-	}
-
-	return u
-}
-
 // valFromObject takes a JSON path as a string and an object of type `unstructured.Unstructured`.
 // This function returns an object of type `reflect.Value` at that JSON path.
-func valFromObject(t *testing.T, jp string, obj *unstructured.Unstructured) reflect.Value {
+func valFromObject(jp string, obj *unstructured.Unstructured) (reflect.Value, error) {
 	jPath := jsonpath.New("parse")
 	if err := jPath.Parse(jp); err != nil {
-		t.Fatalf("Parsing JSONPath: %v", err)
+		return reflect.Value{}, fmt.Errorf("parsing JSONPath: %w", err)
 	}
 
 	v, err := jPath.FindResults(obj.Object)
 	if err != nil {
-		t.Fatalf("Finding results using JSONPath in the YAML file: %v", err)
+		return reflect.Value{}, fmt.Errorf("finding results using JSONPath in the YAML file: %w", err)
 	}
 
 	if len(v) == 0 || len(v[0]) == 0 {
-		t.Fatalf("No result found")
+		return reflect.Value{}, fmt.Errorf("no result found")
```

Comment:
Rather than returning a error here, how about we change the return type to `*reflect.Value` and we `return nil, nil`.

This way, we know everything went well, including query parsing, but the value was simply not found. This will make checking more idiomatic.

Reply:
It is reworked now. The functionality is not changed just returning an error to bubble up the caller stack.

Reply:
But we still must use `strings.Contains`, which IMO should be used as a last resort tool, as it test specific implementation.

With approach I suggested, we avoid ambiguity and we work against the specification of function.

Reply:
The newly added test helper `JSONPathExists` is currently relying on the error from this function call:

```go
v, err := jPath.FindResults(obj.Object)
```

So I don't understand how is the change to the pointer going to help in anything? We will still have to check for the error and match the string.

The error is not coming from this block:

```go
if len(v) == 0 || len(v[0]) == 0 {
	return reflect.Value{}, fmt.Errorf("no result found")
}
```

Also `jPath.FindResults` returns `[][]reflect.Value` which definitely not a array of pointers so we don't need to worry about deep copying it if we convert it to pointer.

Reply:
@surajssd I think calling `jPath.AllowMissingKeys()` should help then?

Reply:
What's wrong with getting an error on something that does not exist?

It is possible that user (in this case a developer) might be giving wrong jsonpath and expecting something to exist. Such cases are ignored with `AllowMissingKeys` or gives wrong results.

Reply:
developer would expect certain value, so they must check for nil values. So either we have const error or return nil, nil, both have similar result with slightly different interface. IMO const error is cleaner from API point of view, but it brings more boilerplate for maintenance. I'm not sure which one is better, other than returning nil is simpler.

Reply:
`AllowMissingKeys` gives us strong gaurantee on which we rely on, so not making at false. But yes returning nil instead of error, then it is up to the user to figure out if there is an empty result.